// ngx-roslib/bundles/ngx-roslib.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs/webSocket'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ngx-roslib', ['exports', '@angular/core', 'rxjs/webSocket', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-roslib'] = {}, global.ng.core, global.rxjs.webSocket, global.rxjs, global.rxjs.operators));
}(this, (function (exports, i0, webSocket, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var NgxRoslibService = /** @class */ (function () {
        function NgxRoslibService() {
            this.isConnected = false;
        }
        Object.defineProperty(NgxRoslibService.prototype, "statusLevel", {
            set: function (level) {
                if (this.ros)
                    this.ros.statusLevel = level;
            },
            enumerable: false,
            configurable: true
        });
        NgxRoslibService.prototype.connect = function (url, transportLibrary, transportOptions) {
            if (transportLibrary === void 0) { transportLibrary = 'websocket'; }
            var ros = (this.ros = new Rosbridge());
            this.connection$ = ros.connect(url, transportLibrary, transportOptions);
            this.onOpen = ros.onOpen;
            this.onClose = ros.onClose;
            this.onError = ros.onError;
            return ros;
        };
        NgxRoslibService.prototype.ngOnDestroy = function () {
            var _a;
            (_a = this.connection$) === null || _a === void 0 ? void 0 : _a.complete();
        };
        return NgxRoslibService;
    }());
    NgxRoslibService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.1", ngImport: i0__namespace, type: NgxRoslibService, deps: [], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    NgxRoslibService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.1", ngImport: i0__namespace, type: NgxRoslibService, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.1", ngImport: i0__namespace, type: NgxRoslibService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root',
                    }]
            }], ctorParameters: function () { return []; } });
    var Rosbridge = /** @class */ (function () {
        function Rosbridge(url, transportLibrary, transportOptions) {
            this.isConnected = false;
            this.transportLibrary = 'websocket';
            this.connectionOpened$ = new rxjs.Subject();
            this.onOpen = this.connectionOpened$.asObservable();
            this.connectionClosed$ = new rxjs.Subject();
            this.onClose = this.connectionClosed$.asObservable();
            this.connectionError$ = new rxjs.Subject();
            this.onError = this.connectionError$.asObservable();
            if (url) {
                this.connection$ = this.connect(url, transportLibrary !== null && transportLibrary !== void 0 ? transportLibrary : 'websocket', transportOptions);
            }
        }
        Object.defineProperty(Rosbridge.prototype, "statusLevel", {
            set: function (level) {
                var setStateLevelRequest = new SetStatusLevel(level);
                this.sendRequest(setStateLevelRequest);
            },
            enumerable: false,
            configurable: true
        });
        Rosbridge.prototype.connect = function (url, transportLibrary, transportOptions) {
            var _this = this;
            if (transportLibrary === void 0) { transportLibrary = 'websocket'; }
            var _a;
            this.transportOptions = transportOptions;
            this.transportLibrary = transportLibrary;
            var wsUrl;
            if (transportLibrary === 'websocket') {
                wsUrl = url.replace(/(http)(s)?:\/\//, 'ws$2://');
                if (!this.connection$ || this.connection$.closed) {
                    this.connection$ = webSocket.webSocket({
                        url: wsUrl,
                        binaryType: 'arraybuffer',
                        openObserver: {
                            next: function (value) {
                                _this.isConnected = true;
                                _this.connectionOpened$.next(value);
                            },
                        },
                        closeObserver: {
                            next: function (closeEvent) {
                                _this.isConnected = false;
                                _this.connectionClosed$.next(closeEvent);
                            },
                        },
                        closingObserver: {
                            next: function () { },
                        },
                    });
                }
            }
            else if (transportLibrary === 'workerSocket') {
                // TODO: add websocket in web worker implementation for the connect method
                //
            }
            (_a = this.connection$) === null || _a === void 0 ? void 0 : _a.subscribe(function () { }, function () { }); // Needed to open the websocket
            return this.connection$;
        };
        Rosbridge.prototype.authenticate = function (mac, client, dest, rand, t, level, end) {
            this.sendRequest({
                op: 'auth',
                mac: mac,
                client: client,
                dest: dest,
                rand: rand,
                t: t,
                level: level,
                end: end,
            });
        };
        Rosbridge.prototype.sendRequest = function (message) {
            var _this = this;
            var _a, _b;
            if (!this.isConnected) {
                console.log('Trying to send request, but not connected. Waiting...');
                (_a = this.onOpen) === null || _a === void 0 ? void 0 : _a.pipe(operators.take(1)).subscribe(function () { var _a; return (_a = _this.connection$) === null || _a === void 0 ? void 0 : _a.next(message); });
            }
            else {
                (_b = this.connection$) === null || _b === void 0 ? void 0 : _b.next(message);
            }
        };
        Rosbridge.prototype.getTopics = function (callback, failedCallback) {
            var topicsService = new RosService({
                ros: this,
                name: '/rosapi/topics',
                serviceType: 'rosapi/Topics',
            });
            if (failedCallback) {
                topicsService.call({}, function (msg) {
                    callback(msg.topics);
                }, failedCallback);
            }
            else {
                topicsService.call({}, function (msg) {
                    callback(msg.topics);
                });
            }
        };
        Rosbridge.prototype.getNodes = function (callback, failedCallback) {
            var topicsService = new RosService({
                ros: this,
                name: '/rosapi/nodes',
                serviceType: 'rosapi/Nodes',
            });
            if (failedCallback) {
                topicsService.call({}, function (msg) {
                    callback(msg.nodes);
                }, failedCallback);
            }
            else {
                topicsService.call({}, function (msg) {
                    callback(msg.nodes);
                });
            }
        };
        return Rosbridge;
    }());
    var idCounter = 0;
    var RosbridgeProtocol = /** @class */ (function () {
        function RosbridgeProtocol() {
        }
        return RosbridgeProtocol;
    }());
    var SetStatusLevel = /** @class */ (function (_super) {
        __extends(SetStatusLevel, _super);
        function SetStatusLevel(level, id) {
            var _this = _super.call(this) || this;
            _this.op = 'set_level';
            _this.level = level;
            _this.id = id;
            return _this;
        }
        return SetStatusLevel;
    }(RosbridgeProtocol));
    var StatusMessage = /** @class */ (function (_super) {
        __extends(StatusMessage, _super);
        function StatusMessage(level, msg, id) {
            var _this = _super.call(this) || this;
            _this.op = 'status';
            _this.level = level;
            _this.id = id;
            _this.msg = msg;
            return _this;
        }
        return StatusMessage;
    }(RosbridgeProtocol));
    var Authenticate = /** @class */ (function (_super) {
        __extends(Authenticate, _super);
        function Authenticate(mac, client, dest, rand, t, level, end) {
            var _this = _super.call(this) || this;
            _this.op = 'auth';
            _this.mac = mac;
            _this.client = client;
            _this.dest = dest;
            _this.rand = rand;
            _this.t = t;
            _this.level = level;
            _this.end = end;
            return _this;
        }
        return Authenticate;
    }(RosbridgeProtocol));
    var Advertise = /** @class */ (function (_super) {
        __extends(Advertise, _super);
        function Advertise(topic, type, id, latch, queue_size) {
            var _this = _super.call(this) || this;
            _this.op = 'advertise';
            _this.topic = topic;
            _this.id = "advertise:" + _this.topic + ":" + ++idCounter;
            _this.type = type;
            _this.latch = latch !== null && latch !== void 0 ? latch : false;
            if (queue_size) {
                _this.queue_size = Math.max(0, queue_size);
            }
            else {
                _this.queue_size = 100;
            }
            return _this;
        }
        return Advertise;
    }(RosbridgeProtocol));
    var Unadvertise = /** @class */ (function (_super) {
        __extends(Unadvertise, _super);
        function Unadvertise(topic, id) {
            var _this = _super.call(this) || this;
            _this.op = 'unadvertise';
            _this.id = id;
            _this.topic = topic;
            return _this;
        }
        return Unadvertise;
    }(RosbridgeProtocol));
    var Publish = /** @class */ (function (_super) {
        __extends(Publish, _super);
        function Publish(topic, msg, id, latch, queue_size) {
            var _this = _super.call(this) || this;
            _this.op = 'publish';
            _this.topic = topic;
            _this.id = "publish:" + _this.topic + ":" + ++idCounter;
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            _this.msg = msg;
            _this.latch = latch !== null && latch !== void 0 ? latch : false;
            if (queue_size) {
                _this.queue_size = Math.max(0, queue_size);
            }
            else {
                _this.queue_size = 100;
            }
            return _this;
        }
        return Publish;
    }(RosbridgeProtocol));
    var Subscribe = /** @class */ (function (_super) {
        __extends(Subscribe, _super);
        function Subscribe(topic, type, throttle_rate, queue_length, fragment_size, compression, id) {
            var _this = _super.call(this) || this;
            _this.op = 'subscribe';
            _this.topic = topic;
            _this.id = "subscribe:" + _this.topic + ":" + ++idCounter;
            _this.type = type;
            _this.throttle_rate = throttle_rate !== null && throttle_rate !== void 0 ? throttle_rate : 0;
            _this.queue_length = queue_length !== null && queue_length !== void 0 ? queue_length : 0;
            _this.fragment_size = fragment_size;
            _this.compression = compression !== null && compression !== void 0 ? compression : 'none';
            return _this;
        }
        return Subscribe;
    }(RosbridgeProtocol));
    var ServiceCall = /** @class */ (function (_super) {
        __extends(ServiceCall, _super);
        function ServiceCall(service, args, id, fragment_size, compression) {
            var _this = _super.call(this) || this;
            _this.args = {};
            _this.op = 'call_service';
            _this.service = service;
            _this.id = id !== null && id !== void 0 ? id : "call_service:" + _this.service + ":" + ++idCounter;
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            _this.args = args !== null && args !== void 0 ? args : {};
            _this.fragment_size = fragment_size;
            _this.compression = compression;
            return _this;
        }
        return ServiceCall;
    }(RosbridgeProtocol));
    var AdvertiseService = /** @class */ (function (_super) {
        __extends(AdvertiseService, _super);
        function AdvertiseService(type, service) {
            var _this = _super.call(this) || this;
            _this.op = 'advertise_service';
            _this.type = type;
            _this.service = service;
            return _this;
        }
        return AdvertiseService;
    }(RosbridgeProtocol));
    var UnadvertiseService = /** @class */ (function (_super) {
        __extends(UnadvertiseService, _super);
        function UnadvertiseService(service) {
            var _this = _super.call(this) || this;
            _this.op = 'unadvertise_service';
            _this.service = service;
            return _this;
        }
        return UnadvertiseService;
    }(RosbridgeProtocol));
    var ServiceRequest = /** @class */ (function (_super) {
        __extends(ServiceRequest, _super);
        function ServiceRequest() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ServiceRequest;
    }(RosbridgeProtocol));
    var ServiceResponse = /** @class */ (function (_super) {
        __extends(ServiceResponse, _super);
        function ServiceResponse(service, result, values, id) {
            var _this = _super.call(this) || this;
            _this.op = 'service_response';
            _this.id = id;
            _this.service = service;
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            _this.values = values !== null && values !== void 0 ? values : {};
            _this.result = result;
            return _this;
        }
        return ServiceResponse;
    }(RosbridgeProtocol));
    var Unsubscribe = /** @class */ (function (_super) {
        __extends(Unsubscribe, _super);
        function Unsubscribe(topic, id) {
            var _this = _super.call(this) || this;
            _this.op = 'unsubscribe';
            _this.topic = topic;
            _this.id = id;
            return _this;
        }
        return Unsubscribe;
    }(RosbridgeProtocol));
    /**
     * RosTopic class used to instantiate a topic, subscribe to its event and advertise future events regarding the topic
     */
    var RosTopic = /** @class */ (function () {
        function RosTopic(_c) {
            var ros = _c.ros, name = _c.name, messageType = _c.messageType, _d = _c.compression, compression = _d === void 0 ? 'none' : _d, _e = _c.throttleRate, throttleRate = _e === void 0 ? 0 : _e, _f = _c.latch, latch = _f === void 0 ? false : _f, _g = _c.queueSize, queueSize = _g === void 0 ? 100 : _g, _h = _c.queueLength, queueLength = _h === void 0 ? 0 : _h, _j = _c.reconnectOnClose, reconnectOnClose = _j === void 0 ? true : _j;
            this.ros = ros;
            this.name = name;
            this.messageType = messageType;
            this.compression = compression;
            this.throttleRate = throttleRate;
            this.latch = latch;
            this.queueSize = queueSize;
            this.queueLength = queueLength;
            this.reconnectOnClose = reconnectOnClose;
            this.throttleRate = Math.max(this.throttleRate, 0);
        }
        RosTopic.prototype.subscribe = function (callback) {
            var _this = this;
            var _a, _b;
            var subRequest = new Subscribe(this.name, this.messageType);
            this.observable$ = (_a = this.ros.connection$) === null || _a === void 0 ? void 0 : _a.pipe(operators.filter(function (data) { return (data === null || data === void 0 ? void 0 : data.topic) === _this.name && (data === null || data === void 0 ? void 0 : data.op) === 'publish'; } // eslint-disable-line @typescript-eslint/no-unsafe-member-access
            ), operators.pluck('msg'));
            this.sub = (_b = this.observable$) === null || _b === void 0 ? void 0 : _b.subscribe(function (message) { return callback(message); });
            this.id = subRequest.id;
            this.ros.sendRequest(subRequest);
        };
        RosTopic.prototype.unsubscribe = function () {
            var _a;
            var unsubRequest = new Unsubscribe(this.name, this.id);
            this.ros.sendRequest(unsubRequest);
            (_a = this.sub) === null || _a === void 0 ? void 0 : _a.unsubscribe();
        };
        RosTopic.prototype.advertise = function () {
            var advertiseRequest = new Advertise(this.name, this.messageType, this.id, this.latch, this.queueSize);
            this.id = advertiseRequest.id;
            this.ros.sendRequest(advertiseRequest);
        };
        RosTopic.prototype.publish = function (msg) {
            var publishRequest = new Publish(this.name, msg, this.id, this.latch, this.queueSize);
            this.ros.sendRequest(publishRequest);
        };
        RosTopic.prototype.unadvertise = function () {
            var unadvertiseRequest = new Unadvertise(this.name, this.id);
            this.ros.sendRequest(unadvertiseRequest);
        };
        return RosTopic;
    }());
    var RosService = /** @class */ (function () {
        function RosService(_c) {
            var ros = _c.ros, name = _c.name, serviceType = _c.serviceType;
            this.isAdvertised = false;
            this.ros = ros;
            this.name = name;
            this.serviceType = serviceType;
        }
        RosService.prototype.advertise = function (callback) {
            var _this = this;
            var _a;
            var serviceAdvertiseRequest = new AdvertiseService(this.serviceType, this.name);
            (_a = this.ros.connection$) === null || _a === void 0 ? void 0 : _a.pipe(operators.filter(function (data) { return (data === null || data === void 0 ? void 0 : data.service) === _this.name && // eslint-disable-line @typescript-eslint/no-unsafe-member-access
                (data === null || data === void 0 ? void 0 : data.op) === 'call_service'; } // eslint-disable-line @typescript-eslint/no-unsafe-member-access
            ), 
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            operators.map(function (_c) {
                var args = _c.args, id = _c.id;
                return ({ args: args, id: id });
            })).subscribe(function (req) {
                var _a;
                var res = callback(req.args);
                var serviceResponse = new ServiceResponse(_this.name, true, res, (_a = req.id) !== null && _a !== void 0 ? _a : '');
                _this.ros.sendRequest(serviceResponse);
            });
            this.ros.sendRequest(serviceAdvertiseRequest);
            this.isAdvertised = true;
        };
        RosService.prototype.call = function (req, callback, failedCallback) {
            var _this = this;
            var _a;
            var serviceCallRequest = new ServiceCall(this.name, req);
            (_a = this.ros.connection$) === null || _a === void 0 ? void 0 : _a.pipe(operators.filter(function (data) { return (data === null || data === void 0 ? void 0 : data.service) === _this.name && // eslint-disable-line @typescript-eslint/no-unsafe-member-access
                (data === null || data === void 0 ? void 0 : data.id) === serviceCallRequest.id && // eslint-disable-line @typescript-eslint/no-unsafe-member-access
                (data === null || data === void 0 ? void 0 : data.op) === 'service_response'; } // eslint-disable-line @typescript-eslint/no-unsafe-member-access
            ), operators.tap(function (msg) {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                if ((msg === null || msg === void 0 ? void 0 : msg.result) !== undefined && (msg === null || msg === void 0 ? void 0 : msg.result) === false)
                    if (failedCallback) {
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                        failedCallback(msg.values);
                    }
            }), 
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            operators.filter(function (data) { return (data === null || data === void 0 ? void 0 : data.result) !== false; }), operators.pluck('values'), operators.take(1)).subscribe(function (message) { return callback(message); });
            this.ros.sendRequest(serviceCallRequest);
        };
        RosService.prototype.unadvertise = function () {
            if (!this.isAdvertised)
                return;
            var serviceUnadvertiseRequest = new UnadvertiseService(this.name);
            this.ros.sendRequest(serviceUnadvertiseRequest);
        };
        return RosService;
    }());
    var RosParam = /** @class */ (function () {
        function RosParam(_c) {
            var ros = _c.ros, name = _c.name;
            this.ros = ros;
            this.name = name;
        }
        RosParam.prototype.get = function (callback) {
            var paramService = new RosService({
                ros: this.ros,
                name: '/rosapi/get_param',
                serviceType: 'rosapi/GetParam',
            });
            paramService.call({ name: this.name }, function (res) {
                // @ts-ignore
                callback(JSON.parse(res.value));
            });
        };
        RosParam.prototype.set = function (newValue, callback) {
            var paramService = new RosService({
                ros: this.ros,
                name: '/rosapi/set_param',
                serviceType: 'rosapi/SetParam',
            });
            paramService.call({ name: this.name, value: JSON.stringify(newValue) }, function () {
                callback();
            });
        };
        RosParam.prototype.delete = function (callback) {
            var paramService = new RosService({
                ros: this.ros,
                name: '/rosapi/delete_param',
                serviceType: 'rosapi/DeleteParam',
            });
            paramService.call({ name: this.name }, function () {
                callback();
            });
        };
        return RosParam;
    }());

    exports.RosoutLevel = void 0;
    (function (RosoutLevel) {
        RosoutLevel[RosoutLevel["DEBUG"] = 1] = "DEBUG";
        RosoutLevel[RosoutLevel["INFO"] = 2] = "INFO";
        RosoutLevel[RosoutLevel["WARN"] = 4] = "WARN";
        RosoutLevel[RosoutLevel["ERROR"] = 8] = "ERROR";
        RosoutLevel[RosoutLevel["FATAL"] = 16] = "FATAL";
    })(exports.RosoutLevel || (exports.RosoutLevel = {}));

    /*
     * Public API Surface of ngx-roslib
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgxRoslibService = NgxRoslibService;
    exports.RosParam = RosParam;
    exports.RosService = RosService;
    exports.RosTopic = RosTopic;
    exports.Rosbridge = Rosbridge;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
